### PR TITLE
Use of DCD for "Go To Definition" and Code Completion functionality plus other changes.

### LINKED ISSUE #5: Go to definition feature is needed

### LINKED ISSUE #20: code completion using DCD
somewhat related to issue #5...

see forum for reference: http://forum.dlang.org/post/wzsoaeqdnftlwwspdazm@forum.dlang.org

maybe coedit's way of using DCD for code completion is an inspiration:
https://github.com/BBasile/Coedit/blob/master/src/ce_dcd.pas#L215

important to note is, that since DCD parses the code itself one needs to save unsaved changes in the file in a tmpfile to have DCD parse it, which should not be a problem and be significantly easier and safer than integrating libdparse yourself. SOA for the win ;)

### PR DESCRIPTION
**Pull request by Freakazo**
Okay so this is all the changes I've made to fix #20 and fix #5.

There is still a fair bit to do, but I believe this is ready to be merged in.

So firstly I've created a new directory for tools, such as Go to Def and Code Completion. In the hope that the IDE in the future can also be used for other languages, I've created the EditorTool class which will contain abstract methods. We then implement those abstract methods for each language inside a folder with the name of the language. DEditorTool implements those methods and uses DCDInterface to start dcd-client to send the requests to dcd-server. I want to re-implement dcd-client so that we can talk directly to dcd-server through sockets, which should be a big boon to performance for very large documents. However I'm completely unsure of the legality of doing so. 
I've also moved SimpleDSyntaxHighlighter out of dsourceedit.d into its own file under tools.

Right now I've assigned Go to Def as SHIFT+G and for Code Completion CTRL+SHIFT+G.
Make sure that dcd-client is in your path and that dcd-server is running.

There still is a fairly lengthy todo list, which I'll get to work on as soon as I get feedback for the changes.

Still to do:
~~Fix caret disappearing after insertion.~~
Refactor extprocess so that we have access to stdin.
Handle the user continueing to type, display only matches containing entered text.
Inform user when dcd-client command is not found.
Check that dcd-server is running and attempt to start it if it is not, if we started it we should also kill it.
Let the user know when we don't know of any completions or we don't know where the symbol declaration is.

I've only tested this on Linux at the moment.

Please review and let me know what I should change.

### CODE REVIEW

File: src/dlangide/ui/dsourceedit.d [hunk 1]
```diff
@@ -245,7 +271,7 @@ class SimpleDSyntaxHighlighter : SyntaxHighlighter {
         for (;;) {
             ch = nextBracket(dir, p);
             if (!ch) // no more brackets
-                return startPos;
+                break;
```
Comment (Freakazo) [inline src/dlangide/ui/dsourceedit.d]: This caused a compiler issue for me at the last return in the function: "Statement is not reachable".

**Comment by buggins** [comment]
automatic merge is impossible

**Comment by buggins** [comment]
Could you please merge to latest mainstream?

**Comment by buggins** [comment]
Merged. It compiles and runs, but I didn't check new functionality

**Comment by buggins** [comment]
Did some refactoring.
Fixed dcd-client running on windows

**Pull request merged**